pkg/core/blockchain.go

```diff
@@ -741,37 +862,208 @@ func (bc *Blockchain) removeOldTransfers(index uint32) time.Duration {
+		nextOlderLog            *storage.KeyValue
+		trimmed                 bool
+		newestTransferTimestamp uint64
+	}
+	var (
+		preEdge  func(batchAcc util.Uint160, k, v []byte, isEdge bool) bool
+		postEdge func(batchAcc util.Uint160, k, v []byte) bool
+		edgeLogs = make(map[util.Uint160]*kvTrimmed)
+	)
+	if removeOld {
+		preEdge = func(_ util.Uint160, _, _ []byte, _ bool) bool {
+			kept++
+			return true
+		}
+		postEdge = func(_ util.Uint160, _, _ []byte) bool {
+			removed++
+			return false
+		}
+	} else {
+		preEdge = func(batchAcc util.Uint160, k, v []byte, isEdge bool) bool {
+			if isEdge {
+				var (
+					oldBatchSize            = v[0]
+					newBatchSize            byte
+					newBatchNewestTimestamp uint64
+				)
+				r := io.NewBinReaderFromBuf(v[1:])
+				l := len(v)
+				bytesRead := 1                            // 1 is for batch size byte which is read by default
+				for i := byte(0); i < oldBatchSize; i++ { // from oldest to newest transfers of the batch
+					var nep17 *state.NEP17Transfer
+					switch k[0] {
+					case byte(storage.STNEP11Transfers):
+						tr := new(state.NEP11Transfer)
+						tr.DecodeBinary(r)
+						if r.Err != nil {
+							// TODO
+							panic(err)
+						}
+						nep17 = &tr.NEP17Transfer
+					case byte(storage.STNEP17Transfers):
+						tr := new(state.NEP17Transfer)
+						tr.DecodeBinary(r)
+						if r.Err != nil {
+							// TODO
+							panic(err)
+						}
+						nep17 = tr
+					}
 
-	for i := range prefixes {
+					if nep17.Block > index {
+						break
+					}
+					bytesRead = l - r.Len() // including batch size byte
+					newBatchNewestTimestamp = nep17.Timestamp
+					newBatchSize++
+				}
+				newLogV := make([]byte, bytesRead)
+				copy(newLogV, v[:bytesRead])
+				newLogV[0] = newBatchSize
+				newLogK := make([]byte, len(k))
+				copy(newLogK, k)
+				edgeLogs[batchAcc] = &kvTrimmed{
+					kv: &storage.KeyValue{
+						Key:   newLogK,
+						Value: newLogV,
+					},
+					newestTransferTimestamp: newBatchNewestTimestamp,
+					nextOlderLog:            nil,
+					trimmed:                 oldBatchSize != newBatchSize,
+				}
+			} else {
+				// Mark this set of transfers for batchAcc as "completely removed" until the edge is reached.
+				edgeLogs[batchAcc] = &kvTrimmed{
+					kv: nil,
+				}
+			}
+			// If it's an edge batch, then remove it anyway and add it to cache.
+			// Relevant transfers of this batch will be re-added to the store later.
+			removed++
+			return false
+		}
+		postEdge = func(batchAcc util.Uint160, k, v []byte) bool {
+			edge := edgeLogs[batchAcc]
+			if edge.nextOlderLog == nil {
+				edge.nextOlderLog = &storage.KeyValue{
+					Key:   k,
+					Value: v,
+				}
+			}
+			kept++
+			return true
+		}
+	}
+	transferInfoCache := make(map[util.Uint160]*state.TokenTransferInfo)
+	for i, p := range prefixes {
 		var acc util.Uint160
-		var canDrop bool
+		var edgePassed bool
+		prefix := prefixes[i : i+1]
 
 		err = bc.store.SeekGC(storage.SeekRange{
```

Comment:
How about `SeekGC(rng SeekRange, keep func(k, v []byte) (bool, []byte))`?

Reply:
It's not needed with the new scheme.